Staging tables

benrudolph opened the pull request:
@gcapalbo first go at an example for group staging. going to have to do some work to get all the source models to have last modified. also a lot of our models have wide ranging data since we've never been too strict about data constraints. i think i can iterate to make better abstractions when i start implementing more of the staging tables. hold off on merge, want to get `br/warehouse` in first

cc: @proteusvacuum

Commit:
group last modified view

Commit:
get_group_ids

Commit:
Intial stage_table command

abc

dimagimon commented on corehq/couchapps/group_last_modified/views/by_last_modified/map.js:
```diff
@@ -0,0 +1,5 @@
+function(doc){
```
Parsing error: Unexpected token (

dimagimon commented on corehq/warehouse/models/__init__.py:
```diff
@@ -10,3 +10,7 @@
     TableState,
 )
+
+from corehq.warehouse.models.staging import (
```
F401 'GroupStagingTable' imported but unused

gcapalbo commented on corehq/warehouse/models/staging.py:
```diff
@@ -0,0 +1,64 @@
+
+class StagingTable(models.Model):
+
+    staged_on = models.DateTimeField(auto_now_add=True)
```
just curious where this value would end up? i think we'd want to include batch id on this table as well

benrudolph replied:
this is actually inherited by all tables that are staging tables. so for each group record in the staging table it will also include the date that it was added. just thought it might be useful meta data. will add a column for batch id, but probably won't address it in this pr until i sort out how the batch state will manifest

gcapalbo replied:
ok, i think batch id might replace this data element since we would know when the batch ran

gcapalbo commented on corehq/warehouse/models/staging.py:
```diff
@@ -0,0 +1,64 @@
+
+    @classmethod
+    def clear_records(cls):
+        cls.objects.all().delete()
```
i think ideally we'd want to make this a truncate so it's faster, but that also means using raw sql

benrudolph replied:
https://github.com/dimagi/commcare-hq/pull/16204/commits/d9185d2d5b111d74692d96bacf0f87d6cf812281

gcapalbo commented on corehq/warehouse/models/staging.py:
```diff
@@ -0,0 +1,64 @@
+                reporting=raw_record.get('reporting'),
+                group_last_modified=raw_record.get('last_modified'),
+            ))
+        cls.objects.bulk_create(records)
```
this is a good idea. for the bigger dims we may want to do 1000 inserts at a time or something like that

benrudolph replied:
yep that sounds right. i was reading this post too: http://stefano.dissegna.me/django-pg-bulk-insert.html. purposely didn't do anything too crazy, but may be worth looking into some of the ideas proposed in that article (using postgres `COPY` or using raw SQL) in addition to batching

gcapalbo commented on corehq/warehouse/management/commands/stage_table.py:
```diff
@@ -0,0 +1,60 @@
+        end = options.get('end')
+
+        model = SLUG_TO_STAGING_TABLE.get(slug)
+        model.stage_records(start, end)
```
i forgot to mention this in our previous chats but it's a best practice any time we're moving data from one step to another to do a balancing step at the end where we make sure the count of records in the source matches the count of records in the destination, and if that doesn't match we should fail the process

benrudolph replied:
👍 good call. in some of the example etl setups i've seen while googling, they have an entire task in the progress called data validation, that does checks like that to ensure that the staged data is ready to be moved to the production tables. probably a bit overkill right now, but something to consider

Commit:
Use truncate

gcapalbo commented:
looks like `test_get_group_ids_by_last_modified` is failing, maybe because other tests don't clean up the groups they create

Commit:
destroy groups on tearDown

Commit:
Remove staged_on

Commit:
extract group staging table

Commit:
group_last_modified to last_modifed for reuse

Commit:
Error message on wrong slug

Commit:
move dbaccessors to warehouse app

Commit:
Add domain staging table

Commit:
bulk insert 1000 at a time

gcapalbo commented:
one thing that just occurred to me is we'll need to figure out how to deal with deleted data. we might want to maintain deleted groups, for example, with a deleted flag

benrudolph commented:
we could also just make it another staging table and iterate through all `Group-Deleted` items (`X-Deleted`) and then update the dim accordingly.

deleted data is going to be much harder to deal with for case or form deletes if we've aggregated the data for fact tables...

Commit:
test group staging table

Commit:
tests for domain staging table

dimagimon commented on corehq/couchapps/last_modified/views/by_last_modified/map.js:
```diff
@@ -0,0 +1,7 @@
+function(doc){
```
Parsing error: Unexpected token (

dimagimon commented on corehq/apps/groups/tests/test_groups.py:
```diff
@@ -1,3 +1,4 @@
+from datetime import datetime, timedelta
```
F401 'datetime' imported but unused
F401 'timedelta' imported but unused

dimagimon commented on corehq/warehouse/models/__init__.py:
```diff
@@ -10,3 +10,8 @@
     TableState,
 )
+
+from corehq.warehouse.models.staging import (
```
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

dimagimon replied:
F401 'DomainStagingTable' imported but unused
F401 'GroupStagingTable' imported but unused

benrudolph commented:
Added the domain staging table and added some more tests and a few more abstractions. i think g2g once we are happy with it. enough for now, going to look at making an `update_warehouse_table` that pulls from the staging tables that are necessary, and upserts them into the dim/fact table

Commit:
dimagimon

gcapalbo commented on corehq/warehouse/models/staging.py:
```diff
@@ -15,12 +16,16 @@ class Meta:
+    @transaction.atomic
+    def stage_records(cls, start_dateime, end_datetime):
+        cls.clear_records()
+        record_iter = cls.raw_record_iter(start_dateime, end_datetime)
```
`start_datetime`

gcapalbo commented on corehq/warehouse/models/staging.py:
```diff
@@ -40,21 +45,15 @@ class GroupStagingTable(StagingTable):
-        cls.objects.bulk_create(records)
+    # Map source model fields to staging table fields
+    # ( <source field>, <staging field> )
+    FIELD_MAPPING = [
```
it might be a good idea to make this a property or be returned from a classmethod, so that we can put it on the superclass and raise a NotImplementError, just so it's clear what needs to be implemented on a new staging table model

benrudolph replied:
👍 will do. i added it as a required param to the bulk create method method, but i think explicitly requiring it in the abstract class is better

gcapalbo commented on corehq/couchapps/last_modified/views/by_last_modified/map.js:
```diff
@@ -1,5 +1,7 @@
     if (doc.doc_type == "Group") {
         emit(doc.last_modified, null);
+    } else if (doc.doc_type == "Domain") {
+        emit(doc.last_modified, null);
```
we need to emit doc type now right, and filter for it in the staging process

gcapalbo replied:
(i realize this might be working right now because they're in different dbs, but i don't think we want to rely on that long term)

benrudolph replied:
👍 will add users now as well so we can avoid the additional reindex

gcapalbo commented:
i think the right way to handle the deleted data for these two dims would actually be to pull in doc type on the staging table, and then transform it to a boolean deleted flag on the dim later on, that way we pull it all at once and don't have to duplicate tables.

agree it's going to be more complicated for the facts when dealing with deleted forms.

benrudolph commented:
> i think the right way to handle the deleted data for these two dims would actually be to pull in doc type on the staging table, and then transform it to a boolean deleted flag on the dim later on, that way we pull it all at once and don't have to duplicate tables.

couch is annoying in that it'll take multiple queries to get the set that we want, but agree i like the idea of transforming the doc_type

Commit:
Include doc_types in last_modified

Commit:
Add doc_type and refactor filed_mapping

benrudolph commented on corehq/couchapps/last_modified/views/by_last_modified/map.js:
```diff
@@ -1,7 +1,12 @@
+    } else if (doc.doc_type == "WebUser" ||
+               doc.doc_type == "WebUser-Deleted" ||
+               doc.doc_type == "CommCareUser" ||
+               doc.doc_type == "CommCareUser-Deleted") {
```
potentially could use base_doc_type here

gcapalbo replied:
actually we'll have to use base doc, at least for the deleted suffix: https://github.com/dimagi/commcare-hq/blob/8063cd3ff1f889d58a997c00557be092bda8e214/corehq/apps/users/models.py#L1611

benrudolph replied:
ah indeed

Commit:
remove unneeded save

Commit:
use iterator

Commit:
clean up before test :disappointed:

Commit:
use batch_size instead of rolling your own

gcapalbo commented:
just one last comment about base_doc for users, otherwise this is looking good

Commit:
use base_doc

Commit:
clean up from other tests

gcapalbo commented on corehq/couchapps/last_modified/views/by_last_modified/map.js:
```diff
@@ -1,12 +1,9 @@
-               doc.doc_type == "CommCareUser-Deleted") {
+    } else if (doc.doc_type === "Domain" || doc.doc_type === "Domain-Deleted") {
         emit([doc.doc_type, doc.last_modified], null);
+    } else if (doc.base_doc && (doc.base_doc === "CouchUser" || doc.doc_type === "CouchUser-Deleted")) {
```
probably just a typo but we'll want to check doc.base_doc for the deleted type too

benrudolph replied:
:man_facepalming: clearly can't write code when there are no tests. making some for when i create the user staging table

Commit:
doc_type -> base_doc

snopoke commented on corehq/couchapps/last_modified/views/by_last_modified/map.js:
```diff
@@ -0,0 +1,9 @@
+    if (doc.doc_type === "Group" || doc.doc_type === "Group-Deleted") {
+        emit([doc.doc_type, doc.last_modified], null);
+    } else if (doc.doc_type === "Domain" || doc.doc_type === "Domain-Deleted") {
+        emit([doc.doc_type, doc.last_modified], null);
```
Could be simpler to use a switch: https://github.com/dimagi/commcare-hq/blob/db598e5b053c741f753cdb1238500bd4aca7b83a/corehq/couchapps/by_domain_doc_type_date/views/view/map.js#L4-L4

benrudolph replied:
something tells me i'm going to have to revisit this view... i think i'm going to opt to merge this now and when i touch this again, i will update to use a switch. there are a couple things i'd like to work on and PR independently that depend on this PR
